Run stable and beta releases on TaskCluster

jgraham opened the pull request.

jgraham committed:
Support installing non-nightly versions of browsers.

Add a --channel argument to ./wpt install and a --browesr-channel
argument to ./wpt run to enable installing beta and stable versions of
the various browsers in addition to nightly.

jgraham committed:
Enable running beta and stable releases on TaskCluster.

Beta releases are run daily, stable releases weekly, in response to
pushes to the relevant epoch/ branches.

jgraham commented:
@jugglinmike if you happen to already have a staging repo configured to run TaskCluster I would appreciate some testing of this branch. If you don't I can set one up myself.

gsnedders commented:
Also can we document how to set something up to test TC changes?

jugglinmike commented on tools/wpt/browser.py:
```diff
@@ -84,9 +84,27 @@ def platform_string_geckodriver(self):
+            "nightly": "latest",
+        }
+        if channel not in branch:
+            raise ValueError
```
Same as above regarding error message

jugglinmike reviewed (commented):
@jgraham in wpt.fyi, we use the terms "stable" and "experimental" to refer to
various browser releases generically. Internally, we map the latter to Chrome's
"dev" channel and Firefox's "nightly" channel. Taking that approach here would
allow us to expose a much more uniform interface. The new terminology will have
to be documented, but it may be easier to explain what we mean by
"experimental" versus having to interpret input like "Chrome nightly" or
"Firefox dev". What do you think?

> @jugglinmike if you happen to already have a staging repo configured to run
> TaskCluster I would appreciate some testing of this branch. If you don't I
> can set one up myself. 

Sure. I had to apply some of my suggestions to get it to run, but here's a push
to "epochs/daily": https://tools.taskcluster.net/groups/PS2oLaLRRZOl0M94VKHLZQ

jugglinmike commented:
> Also can we document how to set something up to test TC changes?

The `.taskcluster.yml` file currently includes the following:

    # Contributors interested in configurating TaskCluster to run
    # against their fork of WPT should change the `workerType` to
    # "github-worker".
    workerType: wpt-docker-worker

I'm open to suggestion for improving this. Do you think more of an explanation
is necessary? Maybe instruction to enable the TaskCluster GitHub app? Or maybe
we should move this elsewhere--the top of the file, or a dedicated page in the
`docs` directory?

jgraham commented:
Regarding using experimental/stable, I don't think that quite works because we have three states; nightly, beta and stable. In fact for Chrome there are four states; canary, dev, beta and stable. Running stable and beta gives us the ability to test version N+1 and version N simultaneously (assuming that beta and stable are both using the same configuration which is more or less true for gecko at least).

I think that latest/beta/stable would be a sufficient ontology to cover our needs (although I'm not exactly sure how Edge and WebKit/Safari fit in there). In terms of actual command line arguments, if we expect people to run this generally I would prefer to be liberal in what we accept rather than forcing people to learn a specific set of nomenclature (e.g. I would happily accept all of experimetal, nightly, dev, preview, latest to mean whatever the most up to date version of the browser is (dev is perhaps the least appealing since Firefox has a developer edition that isn't quite the same thing). And equally I would happily accept stable, release, and other synonyms used for the stable channel.

jgraham commented on .taskcluster.yml:
```diff
@@ -4,10 +4,19 @@ policy:
+        $map:
+          $if: event.ref == 'refs/heads/master'
+          then:
+            {config: [{name: firefox, channel: nightly}, {name: chrome, channel: dev}]}
```
I didn't do that because it's not exhaustive, but `if`/`else` is. I agree that `$match` is more readable, but I don't know what the right tradeoff is. Maybe I shouldn't care about handling every case since hitting a case that isn't handled here resulting in a runtime error would be fine.

jgraham commented:
https://tools.taskcluster.net/groups/KkY09Cj_QKO4A2JPHUyn_Q is a run starting from a `epochs/daily' push on my fork.

jgraham commented:
@foolip created some infrastructure for pushing to the `epochs` branches; see e.g. https://github.com/web-platform-tests/wpt/tree/epochs/daily

jgraham committed:
Pass browser-channel through to wptrunner.

This allows it to be recorded in the run info data and hence wptreport files.

jgraham committed:
Fix getting prefs for Firefox beta releases.

The channel derived from the binary turns out to not always be reliable.
Also we were using an incorrect tagging scheme for beta prefs, so we tried
to download a file that didn't exist. To prevent clashes store downloaded firefox
and its prefs in channel-specific directories.

jugglinmike commented on .taskcluster.yml:
```diff
@@ -4,10 +4,16 @@ policy:
   $if: tasks_for == "github-push"
   then:
-    $if: event.ref == "refs/heads/master"
+    $if: event.ref == "refs/heads/master" || event.ref == "refs/heads/epochs/daily" || event.ref == "refs/heads/epochs/weekly"
```
With the introduction of `$match`, this condition is no longer necessary.

jgraham committed:
Remove unnecessary if condition from .taskcluster.yml

jgraham committed:
Automatically vary workerType by repository name

jgraham commented:
https://tools.taskcluster.net/groups/NTBwJqX_TNiRKL-FEPegPw is a run of this.

foolip commented:
Exciting! I guess that wasn't triggered by a branch update, but otherwise is the same as what we want?

jgraham commented:
It wasn't but I can do that too if you like. I had an earlier run that was triggered by a branch update so I'm pretty sure that works, I was mostly checking my refactoring didn't break the yml file entirely.

jgraham commented:
https://tools.taskcluster.net/groups/CT4NJclrTG-UKAg8FC0Kvw is a beta run on the daily branch.

foolip commented:
Very cool! @Hexcles, I presume that we're still a ways away from ingesting results into wpt.fyi? Do we have some scripts that'd allow us to do a one-off comparison between that run and something close for stable?

gsnedders reviewed (approved):
I think this LGTM, but I don't exactly know much about TC. :)

gsnedders reviewed (approved):
I think this LGTM, but I don't exactly know much about TC. :)

jugglinmike commented on .taskcluster.yml:
```diff
@@ -4,71 +4,76 @@ policy:
+          created: {$fromNow: ''}
+          deadline: {$fromNow: '24 hours'}
+          provisionerId: aws-provisioner-v1
+          workerType:
```
Good idea!

jugglinmike commented on tools/wpt/install.py:
```diff
@@ -2,20 +2,57 @@
+    'nightly': latest_channels,
+    'dev': latest_channels,
+    'preview': latest_channels,
+    'experimental': latest_channels,
```
The terms "experimental", "release" and "preview" aren't used by either browser. If we're going to support both "beta" and "newer than beta," I'm afraid those may confuse folks. Even if the confusion is rare, it seems better to omit them since they are all aliases for more familiar terms.

Alternatively, we could document their meaning in the `--help` text.

Hexcles replied:
I suppose "preview" might be used for the future Safari support, perhaps?

jgraham commented:
@jugglinmike Added some more help text.

jgraham committed:
Install nightly geckodriver when testing Firefox nightly.

This (ab)uses mozdownload to grab the common.tests.tar.gz package from the
latest nightly build and extracts geckodriver from that. The main disadvantage
is that it's rather slow since the package itself is rather large and we want
only a small part of it. The solution to this would be to package up
geckodriver by itself on taskcluster, but that will require changes in the
gecko buildsystem.

Hexcles commented on tools/wpt/browser.py:
```diff
@@ -202,8 +238,14 @@ def get_profile_bundle_url(self, version, channel):
+        version, channel_ = self.get_version_and_channel(binary)
+        if channel is not None and channel != channel_:
+            # Beta doesn't always seem to have the b in the version string, so allow the
+            # manually supplied value to override the one from the binary
```
Wow, really? This is good to know. It's yet another reason why we need to include channels in `wptreport.json`. Currently the receiver infers the channel from the version string, which according to this comment is apparently unreliable.

Hexcles commented on tools/wptrunner/wptrunner/wptcommandline.py:
```diff
@@ -192,6 +192,12 @@ def create_parser(product_choices=None):
                               default=None, help="Browser against which to run tests")
+    config_group.add_argument("--browser-version", action="store",
+                              default=None, help="Informative string detailing the browser "
+                              "release version. This is included in the run_info data.")
```
In my earlier change, I added this to `wpt/run.py`:

https://github.com/web-platform-tests/wpt/blob/1de1550e95396f979da97620a289c213acaf55e2/tools/wpt/run.py#L439

Would it always override the command line argument?

Hexcles reviewed (commented):
Just got one question regarding the `--browser-version` argument.

jugglinmike committed:
Remove downloaded artifact in all cases

jugglinmike committed:
Merge pull request #12824 from web-platform-tests/geckodriver_nightly

Install nightly geckodriver when testing Firefox nightly.

foolip commented:
Oh cool, this is merged! epochs/daily currently points to https://github.com/web-platform-tests/wpt/commit/9a5d71b326166e12784bdd9d161772e20f87c1fd and I see that we had beta results:
https://tools.taskcluster.net/groups/dMjbBGL-TPu9z6TuZYzonw

However, I can't find the regular (experimental) results. Did beta run *instead* of experimental?

foolip commented on .taskcluster.yml:
```diff
@@ -4,71 +4,76 @@ policy:
+      $map:
+        $flatten:
+          $match: {
+            event.ref == "refs/heads/master": [{name: firefox, channel: nightly}, {name: chrome, channel: dev}],
```
The answer to my question is probably here, but I can't interpret it. Pushes to master and pushes to epochs/* will be separate events, but they will be the same sha. Does Taskcluster do something to avoid retesting the same sha? Or is it perhaps just task group triggered by push to master is now hidden in GitHub's UI? @Hexcles, if that is the case, will your integration code be able to find both after the fact, or will it only work if we react to the webhooks as the runs are happening?

jgraham replied:
The answer is actually at https://api.github.com/repos/web-platform-tests/wpt/commits/9a5d71b326166e12784bdd9d161772e20f87c1fd/statuses There are two statuses with context `Taskcluster (push)`, one with the URL https://tools.taskcluster.net/groups/dMjbBGL-TPu9z6TuZYzonw and one with the URL https://tools.taskcluster.net/groups/FFjitCXpQ2C4fD_7O7gPeA GitHub groups by context so it only displays one push. AFAIK there isn't a way to have the TaskCluster tasks get different GitHub context values (possibly we could add that feature?) so we can still get at all the results but can't see them all in the GH UI.

foolip replied:
Cool, thanks! The epochs/weekly branch currently points to 9614def367ae8af8247142c34b00f755863d2fbb, and [the statuses API](https://api.github.com/repos/web-platform-tests/wpt/commits/9614def367ae8af8247142c34b00f755863d2fbb/statuses) does include three runs: [stable](https://tools.taskcluster.net/groups/Dh4Iu9GTRNaSjRLNZ3Sr9w), [beta](https://tools.taskcluster.net/groups/bXctOOvuQriKNr5T-oKERg) and [dev](https://tools.taskcluster.net/groups/TZSm9Pa8TlaP-CCdaA_KPQ).

@Hexcles, is that the API you use to find Taskcluster runs?

Hexcles replied:
My code listens to the "status" webhook events, whose payload is similar to that API. There'll be two separate updates in this scenario and my code will run twice. Therefore, all Taskcluster runs will be uploaded as long as the status is "success", regardless of how GitHub UI displays them and whether they are on the same revision.

GEEKGIRL36 commented:
Pppppp

On Sep 10, 2018 3:41 AM, "jgraham" <notifications@github.com> wrote:

*@jgraham* commented on this pull request.
------------------------------

In .taskcluster.yml
<https://github.com/web-platform-tests/wpt/pull/12679#discussion_r216275262>
:

> -                public/results:
-                  path: /home/test/artifacts
-                  type: directory
-              command:
-                - /bin/bash
-                - --login
-                - -c
-                - "~/start.sh ${event.repository.url}
${event.ref[len('refs/heads/'):]} ${event.after}
${browser.name}-${browser.channel} &&
-                  cd ~/web-platform-tests &&
-                  ./tools/ci/ci_taskcluster.sh ${browser.name}
${chunk[0]} ${chunk[1]} ${chunk[2]}"
-    else: []
+    $flattenDeep:
+      $map:
+        $flatten:
+          $match: {
+            event.ref == "refs/heads/master": [{name: firefox,
channel: nightly}, {name: chrome, channel: dev}],

The answer is actually at
https://api.github.com/repos/web-platform-tests/wpt/commits/9a5d71b326166e12784bdd9d161772e20f87c1fd/statuses
There are two statuses with context Taskcluster (push), one with the URL
https://tools.taskcluster.net/groups/dMjbBGL-TPu9z6TuZYzonw and one with
the URL https://tools.taskcluster.net/groups/FFjitCXpQ2C4fD_7O7gPeA GitHub
groups by context so it only displays one push. AFAIK there isn't a way to
have the TaskCluster tasks get different GitHub context values (possibly we
could add that feature?) so we can still get at all the results but can't
see them all in the GH UI.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/web-platform-tests/wpt/pull/12679#discussion_r216275262>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/Adl9mgQW2kQKSgA3N17Zvv9YyzaoG3uFks5uZkHCgaJpZM4WLvot>
.

whimboo commented:
@jgraham are we creating artifacts especially for the logs? If not yet, what is the follow-up issue to get this integrated?

jugglinmike commented:
Log files are available for all builds. They can be downloaded via the TaskCluster UI or via the command `./wpt tc-download` (more detail on that is available from the `--help` argument).

whimboo commented:
@jugglinmike I tried but I cannot find any log for https://tools.taskcluster.net/groups/ccZFTpXgQ1CqBIHft81eXQ which was triggered by https://github.com/web-platform-tests/wpt/pull/12968. Taskcluster doesn't list any artifacts nor logs.

whimboo commented:
Oh, wait. That is actually a task group! Navigating to a single test job I can see the artifacts now. So all is fine.

Hexcles commented:
@whimboo https://tools.taskcluster.net/groups/ccZFTpXgQ1CqBIHft81eXQ/tasks/DuJQTji0QomkUiTD3q3CCg/runs/0/logs/public%2Flogs%2Flive.log

You need to select a task in the group first. Then you'll see some logs in the "Run Logs" dropdown.

By the way, this particular example had no "affected tests", so no tests ran.
